packages/runtime/src/components/builtin/Text/ReadOnlyText.tsx

```diff
@@ -0,0 +1,171 @@
+import { cx } from '@emotion/css'
+import { ForwardedRef, forwardRef, ReactNode } from 'react'
+import type { BlockJSON, DocumentJSON, InlineJSON, MarkJSON, NodeJSON, TextJSON } from 'slate'
+import type {
+  ElementIDValue,
+  LinkValue,
+  RichTextValue,
+} from '../../../prop-controllers/descriptors'
+import { useStyle } from '../../../runtimes/react/use-style'
+import { Link } from '../../shared/Link'
+import DeviceOverrideBlock from './components/RichTextEditor/components/Block'
+import DeviceOverrideMark from './components/RichTextEditor/components/Mark'
+
+type Props = {
+  id?: ElementIDValue
+  text?: RichTextValue
+  width?: string
+  margin?: string
+}
+
+const ReadOnlyText = forwardRef(function ReadOnlyText(
+  { id, text, width, margin }: Props,
+  ref: ForwardedRef<HTMLDivElement>,
+) {
+  const plaintext = text?.document == null ? '' : getNodeText(text.document)
+
+  return (
+    <div ref={ref} id={id} className={cx(width, margin)}>
+      {plaintext === '' ? <Placeholder /> : <Node {...text?.document} />}
+    </div>
+  )
+})
```

Comment:
The reader component is simpler than I expected. Did you copy-paste the rendering logic/markup from Slate or is this written from scratch?

I'm just worried that there will be a discrepancy between the reader component and the editable component.

Reply:
I wrote it from scratch. There _is_ a discrepancy. I purposefully kept the DOM structure a bit simpler. I looked at the Slate source code to attempt to make it as close as possible, but honestly, Slate's markup is trash and I want us to handcraft it moving forward, specially once we upgrade Slate and we have more control over that.

As for the difference in markup, I mention this in the changelog and it's also why I marked the change as a breaking one. People shouldn't be relying on the DOM structure of our Text component anyway

Reply:
Okay, nice. The code looks clean so should be easy if want to update it. 👍